python/mxnet/executor_manager.py

```diff
@@ -0,0 +1,341 @@
+                 'arguments are %s, auxiliary params are %s'
+                ) % (name, str(arg_names), str(aux_names)))
+        aux_set.add(name)
+
+def _load_general(data, targets):
+    """Load a list of arrays into a list of arrays specified by slices"""
+    for d_src, d_targets in zip(data, targets):
+        if isinstance(d_targets, nd.NDArray):
+            d_src.copyto(d_targets)
+        else:
+            for slice_idx, d_dst in d_targets:
+                d_src[slice_idx].copyto(d_dst)
+
+def _load_data(batch, targets):
+    """Load data into sliced arrays"""
+    _load_general(batch.data, targets)
+
+def _load_label(batch, targets):
+    """Load label into sliced arrays"""
+    _load_general(batch.label, targets)
+
+def _bind_exec(self, sym, ctx, input_shapes, param_names, need_grad=False,
+              base_exec=None, shared_data_arrays=None):
+    """bind executor for bucketing, potentially sharing data with an existing executor."""
+    arg_shape, _, aux_shape = sym.infer_shape(**input_shapes)
+    arg_arrays = []
+    grad_arrays = {} if need_grad else None
+
+    arg_names = sym.list_arguments()
+
+    # create or borrow arguments and gradients
+    for i in range(len(arg_names)):
+        name = arg_names[i]
+        if not name in param_names:
+            # data or label
+            if shared_data_arrays is not None and \
+                    name in shared_data_arrays:
+                arg_arr = shared_data_arrays[name]
+
+                # in bucketing, we want to be strict here to avoid
+                # potential bugs
+                assert(arg_shape[i] == arg_arr.shape)
+                assert(arg_arr.context == ctx[i])
+            else:
+                arg_arr = nd.zeros(arg_shape[i], ctx)
+            arg_arrays.append(arg_arr)
+            if shared_data_arrays is not None:
+                shared_data_arrays[name] = arg_arr
+        else:
+            # model parameter
+            if base_exec is None:
+                arg_arr = nd.zeros(arg_shape[i], ctx)
+                if need_grad:
+                    grad_arr = nd.zeros(arg_shape[i], ctx)
+                    grad_arrays[name] = grad_arr
+            else:
+                arg_arr = base_exec.arg_arrays[i]
+                assert arg_arr.shape == arg_shape[i]
+                if need_grad:
+                    grad_arrays[name] = base_exec.grad_arrays[i]
+            arg_arrays.append(arg_arr)
+
+    # create or borrow aux variables
+    if base_exec is None:
+        aux_arrays = [nd.zeros(s, ctx) for s in aux_shape]
+    else:
+        aux_arrays = [a for a in base_exec.aux_arrays]
+
+    executor = sym.bind(ctx=ctx, args=arg_arrays, args_grad=grad_arrays,
+                        aux_states=aux_arrays,
+                        grad_req='write' if need_grad else 'null', shared_exec=base_exec)
+    return executor
+
+class DataParallelExecutorGroup(object):
+    def __init__(self, sym, param_names, ctx, slices, train_data, shared_group=None):
```

Comment:
Would it be possible to add some comments to describe what this class was? I had to do a bit of digging to actually figure it out.